airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/thread_based_concurrent_stream.py

```diff
@@ -135,10 +142,17 @@ def _wait_while_too_many_pending_futures(self, futures: List[Future[Any]]) -> No
     def _check_for_errors(self, futures: List[Future[Any]]) -> None:
         exceptions_from_futures = [f for f in [future.exception() for future in futures] if f is not None]
         if exceptions_from_futures:
-            raise RuntimeError(f"Failed reading from stream {self.name} with errors: {exceptions_from_futures}")
-        futures_not_done = [f for f in futures if not f.done()]
-        if futures_not_done:
-            raise RuntimeError(f"Failed reading from stream {self.name} with futures not done: {futures_not_done}")
+            exception = RuntimeError(f"Failed reading from stream {self.name} with errors: {exceptions_from_futures}")
+            self._stop_and_raise_exception(exception)
+        else:
+            futures_not_done = [f for f in futures if not f.done()]
+            if futures_not_done:
+                exception = RuntimeError(f"Failed reading from stream {self.name} with futures not done: {futures_not_done}")
+                self._stop_and_raise_exception(exception)
+
+    def _stop_and_raise_exception(self, exception: BaseException) -> None:
+        self._threadpool.shutdown(wait=False, cancel_futures=True)
```

Comment:
How do you feel about the fact that enqueuing the exception puts it at the end of the queue, so we're still processing a bunch of records/partitions before we actually shut down the threadpool? 

It doesn't seem ideal to me, but I'm not sure how big a deal it is. If we wanted to prioritize the exceptions I think we could just do a pretty clean swap of `Queue` with `PriorityQueue` (with an additional `priority` rating when we submit items).

Reply:
that's a good idea! let's make this a fast follow up